vpr/src/place/simpleRL_move_generator.cpp

```diff
@@ -0,0 +1,317 @@
+        reward /= time_elapsed_[last_action_];
+
+    //Determine step size
+    float step = 0.;
+    if (exp_alpha_ < 0.) {
+        step = 1. / n_[last_action_]; //Incremental average
+    } else if (exp_alpha_ <= 1) {
+        step = exp_alpha_; //Exponentially wieghted average
+    } else {
+        VTR_ASSERT_MSG(false, "Invalid step size");
+    }
+
+    //Based on the outcome how much should our estimate of q change?
+    float delta_q = step * (reward - q_[last_action_]);
+
+    //Update the estimated value of the last action
+    q_[last_action_] += delta_q;
+
+#if 0
+    //Update the cummulative q array
+    for (size_t i = last_action_; i < k_; ++i) {
+        cumm_q_[i] += delta_q;
+    }
+#endif
+
+    if (f_) {
+        fprintf(f_, "%zu,", last_action_);
+        fprintf(f_, "%g,", reward);
+
+        for (size_t i = 0; i < k_; ++i) {
+            fprintf(f_, "%g,", q_[i]);
+        }
+
+        for (size_t i = 0; i < k_; ++i) {
+            fprintf(f_, "%zu", n_[i]);
+            if (i != k_ - 1) {
+                fprintf(f_, ",");
+            }
+        }
+        fprintf(f_, "\n");
+    }
+}
+
+/*                                  *
+ *                                  *
+ *  E-greedy agent implementation   *
+ *                                  *
+ *                                  */
+EpsilonGreedyAgent::EpsilonGreedyAgent(size_t k, float epsilon) {
+    //f_ = vtr::fopen("agent_info.txt", "w");
+    set_epsilon(epsilon);
+    set_k(k);
+    set_epsilon_action_prob();
+}
+
+EpsilonGreedyAgent::~EpsilonGreedyAgent() {
+    if (f_) vtr::fclose(f_);
+}
+
+void EpsilonGreedyAgent::set_step(float gamma, int move_lim) {
+    //VTR_LOG("Setting egreedy step: %g\n", exp_alpha_);
+    if (gamma < 0) {
+        exp_alpha_ = -1; //Use sample average
+    } else {
+        //
+        // For an exponentially wieghted average the fraction of total weight applied to
+        // to moves which occured > K moves ago is:
+        //
+        //      gamma = (1 - alpha)^K
+        //
+        // If we treat K as the number of moves per temperature (move_lim) then gamma
+        // is the fraction of weight applied to moves which occured > move_lim moves ago,
+        // and given a target gamma we can explicitly calcualte the alpha step-size
+        // required by the agent:
+        //
+        //     alpha = 1 - e^(log(gamma) / K)
+        //
+        float alpha = 1 - std::exp(std::log(gamma) / move_lim);
+        exp_alpha_ = alpha;
+    }
+}
+
+size_t EpsilonGreedyAgent::propose_action() {
+    size_t action = 0;
+    if (vtr::frand() < epsilon_) {
+        float p = vtr::frand();
+        auto itr = std::lower_bound(cumm_epsilon_action_prob_.begin(), cumm_epsilon_action_prob_.end(), p);
+        action = itr - cumm_epsilon_action_prob_.begin();
+    } else {
+        //Greedy
+        auto itr = std::max_element(q_.begin(), q_.end());
+        VTR_ASSERT(itr != q_.end());
+        action = itr - q_.begin();
+
+#if 0
```

Comment:
Why is this commented out?

Reply:
That was test code. I have cleaned this file which initially contains many test code snippets. Thanks for pointing that to me.